src/osd/objclass.cc

```diff
@@ -700,3 +700,44 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
 
   return ctx->pg->get_pool().stripe_width;
 }
+
+struct GatherFinisher : public PrimaryLogPG::OpFinisher {
+  std::shared_ptr<std::map<std::string, bufferlist> > src_obj_buffs;
+  GatherFinisher(std::shared_ptr<std::map<std::string, bufferlist> > sob) : src_obj_buffs(sob) {}
+  int execute() override {
+    return 0;
+  }
+};
+
+int cls_cxx_gather(cls_method_context_t hctx, const std::set<std::string> &src_objs, const std::string& pool,
+		   const char *cls, const char *method, bufferlist& inbl)
+{
+  PrimaryLogPG::OpContext **pctx = (PrimaryLogPG::OpContext**)hctx;
+  std::shared_ptr<std::map<std::string, bufferlist> > src_obj_buffs(new std::map<std::string, bufferlist>);
+  for (std::set<std::string>::const_iterator it = src_objs.begin(); it != src_objs.end(); ++it) {
+    (*src_obj_buffs)[*it] = bufferlist();
+  }
+  (*pctx)->op_finishers[(*pctx)->current_osd_subop_num].reset(new GatherFinisher(src_obj_buffs));
+  return (*pctx)->pg->start_cls_gather(*pctx, src_obj_buffs, pool, cls, method, inbl);
+}
+
+int cls_cxx_get_gathered_data(cls_method_context_t hctx, std::map<std::string, bufferlist> *results)
+{
+  PrimaryLogPG::OpContext **pctx = (PrimaryLogPG::OpContext**)hctx;
+  PrimaryLogPG::OpFinisher* op_finisher = nullptr;
+  int r = 0;
+  {
+    auto op_finisher_it = (*pctx)->op_finishers.find((*pctx)->current_osd_subop_num);
+    if (op_finisher_it != (*pctx)->op_finishers.end()) {
+      op_finisher = op_finisher_it->second.get();
+    }
+  }
+  if (op_finisher == NULL) {
+    results->clear();
+  } else {
+    GatherFinisher *gf = (GatherFinisher*)op_finisher;
+    *results = *gf->src_obj_buffs;
```

Comment:
swap?

Reply:
Looks like this hasn't been addressed?

Reply:
Also, guard with a null check on results.  When using a pointer out param like results here, we normally permit a nullptr to indicate that the return should be discarded.